govcd/extension_test.go

```diff
@@ -24,7 +26,112 @@ func (vcd *TestVCD) Test_GetExternalNetwork(check *C) {
 	check.Assert(err, IsNil)
 	LogExternalNetwork(*externalNetwork)
 	check.Assert(externalNetwork.HREF, Not(Equals), "")
-	expectedType := "application/vnd.vmware.admin.extension.network+xml"
 	check.Assert(externalNetwork.Name, Equals, networkName)
-	check.Assert(externalNetwork.Type, Equals, expectedType)
+	check.Assert(externalNetwork.Type, Equals, types.MimeExtensionNetwork)
+}
+
+func (vcd *TestVCD) Test_CreateExternalNetwork(check *C) {
```

Comment:
Can you explain what this race condition is? - cause like from external library comes.

Reply:
It actually gives this error everytime (wonder if it's on my env only).
It wasn't  full output as the full is pretty big.
```
WARNING: DATA RACE                                                                                                                                                                [169/3097]
Read at 0x00c0000bed10 by goroutine 31:
  github.com/kr/pretty.(*printer).printValue()
      /usr/local/Cellar/go/1.12.4/libexec/src/reflect/value.go:952 +0xd5c
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:182 +0x228c
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:249 +0x1963
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:182 +0x228c
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:249 +0x1963
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:228 +0x34f
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:137 +0x14a4
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:182 +0x228c
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:249 +0x1963
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:182 +0x228c
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:182 +0x228c
  github.com/kr/pretty.(*printer).printValue()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:249 +0x1963
  github.com/kr/pretty.formatter.Format()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/formatter.go:57 +0x223
  github.com/kr/pretty.(*formatter).Format()
      <autogenerated>:1 +0xab
  fmt.(*pp).handleMethods()
      /usr/local/Cellar/go/1.12.4/libexec/src/fmt/print.go:583 +0x186
  fmt.(*pp).printArg()
      /usr/local/Cellar/go/1.12.4/libexec/src/fmt/print.go:699 +0x24f
  fmt.(*pp).doPrintf()
      /usr/local/Cellar/go/1.12.4/libexec/src/fmt/print.go:1016 +0x31c
  fmt.Sprintf()
      /usr/local/Cellar/go/1.12.4/libexec/src/fmt/print.go:214 +0x73                                                                                                              [132/3097]
  github.com/kr/pretty.(*sbuf).Printf()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/diff.go:12 +0x79
  github.com/kr/pretty.diffPrinter.diff()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/diff.go:76 +0x2dba
  github.com/kr/pretty.diffPrinter.diff()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/diff.go:170 +0x373c
  github.com/kr/pretty.Pdiff()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/diff.go:44 +0xaf
  gopkg.in/check%2ev1.formatUnequal()
      /Users/dserplis/go/pkg/mod/github.com/kr/pretty@v0.1.0/diff.go:19 +0x1aa
  gopkg.in/check%2ev1.(*equalsChecker).Check()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/checkers.go:238 +0x159
  gopkg.in/check%2ev1.(*notChecker).Check()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/checkers.go:94 +0xa2
  gopkg.in/check%2ev1.(*C).internalCheck()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/helpers.go:214 +0x29e
  gopkg.in/check%2ev1.(*C).Assert()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/helpers.go:173 +0xa6
  github.com/vmware/go-vcloud-director/v2/govcd.(*TestVCD).Test_CreateExternalNetwork()
      /Users/dserplis/Documents/vmware/go-vcloud-director/govcd/extension_test.go:96 +0x10a2
  runtime.call32()
      /usr/local/Cellar/go/1.12.4/libexec/src/runtime/asm_amd64.s:519 +0x3a
  reflect.Value.Call()
      /usr/local/Cellar/go/1.12.4/libexec/src/reflect/value.go:308 +0xc0
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/check.go:781 +0x978
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/check.go:675 +0xb7

Previous write at 0x00c0000bed10 by goroutine 29:
  bufio.(*Reader).Peek()
      /usr/local/Cellar/go/1.12.4/libexec/src/bufio/bufio.go:135 +0x79
  net/http.(*persistConn).readLoop()
      /usr/local/Cellar/go/1.12.4/libexec/src/net/http/transport.go:1677 +0x27b

Goroutine 31 (running) created at:
  gopkg.in/check%2ev1.(*suiteRunner).forkCall()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/check.go:672 +0x44e
  gopkg.in/check%2ev1.(*suiteRunner).forkTest()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/check.go:763 +0x1b9
  gopkg.in/check%2ev1.(*suiteRunner).run()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/check.go:818 +0x226
  gopkg.in/check%2ev1.Run()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/run.go:92 +0x5a
  gopkg.in/check%2ev1.RunAll()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/run.go:84 +0x129
  gopkg.in/check%2ev1.TestingT()
      /Users/dserplis/go/pkg/mod/gopkg.in/check.v1@v1.0.0-20180628173108-788fd7840127/run.go:72 +0x6a4
  github.com/vmware/go-vcloud-director/v2/govcd.Test()
      /Users/dserplis/Documents/vmware/go-vcloud-director/govcd/api_vcd_test.go:233 +0x38
  testing.tRunner()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:865 +0x163

Goroutine 29 (running) created at:
  net/http.(*Transport).dialConn()
      /usr/local/Cellar/go/1.12.4/libexec/src/net/http/transport.go:1357 +0xc32
  net/http.(*Transport).getConn.func4()
      /usr/local/Cellar/go/1.12.4/libexec/src/net/http/transport.go:1015 +0xd0
==================
```
Line 95 triggers it, but don't understand why as the task should have been processed by that time

Reply:
It gives for me too, if I recall when I checked. But don't see option to solve it cause `check.Assert` isn't our thing ^^

Reply:
Just for future reference. We think http.Client (which is embedded into task struct) does it's own cleanups in a goroutine (Goroutine 29 in the example above) and triggers data race because assertion in line 95 is accessing it.